### ISSUE TITLE
Rule proposal: disallow optional chaining on undeclared variables

### ISSUE THREAD

**Issue by fregante** [sindresorhus/eslint-plugin-unicorn#1479]
It's unfortunate that `?.` is sugar for `x !== undefined` and not `typeof x !== 'undefined'`. The subtle difference is that the first one will fail if `x` is not defined at all.

Would it make sense to disallow the operator unless the variable is in the scope?

## Fail

```js
iDontExist?.meNeither
```

## Pass

```js
globalThis.iDontExist?.meNeither
```

```js
let iDontExist;
iDontExist?.meNeither
```

**Comment by sindresorhus** [comment]
> It's unfortunate that ?. is sugar for x !== undefined and not typeof x !== 'undefined'

I view that as a feature. It would have caused a lot of silent failures that are hard to debug. JS is loose enough as it is.

**Comment by sindresorhus** [comment]
> Would it make sense to disallow the operator unless the variable is in the scope?

Makes sense to me.

**Comment by sindresorhus** [comment]
Any suggestion for the rule name?

**Comment by fregante** [comment]
`no-undeclared-optional-chaining`?

**Comment by sindresorhus** [comment]
Alternative: `no-optional-chaining-on-undeclared-variable` (a bit verbose, but more correct)

**Comment by sindresorhus** [comment]
@fisker What's your vote on this rule?

**Comment by fisker** [comment]
I missed this proposal. Sounds useful to me.:+1:

**Comment by sindresorhus** [comment]
Let's go with the naming `no-optional-chaining-on-undeclared-variable`.

**Comment by sindresorhus** [comment]
This is now accepted.

**Comment by fisker** [comment]
`no-undeclared-variable-optional-member`?

**Comment by fisker** [comment]
Ah, it can also be optional call, `no-optional-chaining-on-undeclared-variable` is better, if this rule also check optional call, and I think we should.
